[FAB-17433] Add _lifecycle function for querying approved chaincode

satota2 opened the pull request:
This patch adds a function for querying the details of the chaincode definition approved by the organization itself to `_lifecycle`, so that Fabric peer can provide a function to query them.

#### Type of change

- New feature

#### Description

Currently, Fabric peer does not provide a function to query the details of approved chaincode definitions.

On the other hand, Fabric admins will need to confirm/use the information after approval for their operations.

So, we propose to add a function for querying the details of the approved chaincode definition to Fabric peer CLI (Refer to https://jira.hyperledger.org/browse/FAB-17401).

As a sub-task of FAB-17401, this patch adds a function for querying the details of the chaincode definition approved by the organization itself to `_lifecycle`, so that Fabric peer can provide a function to query them.

The following patch has been merged to extend lifecycle protos:
- https://github.com/hyperledger/fabric-protos/pull/25

Immediately after this patch has been merged into the fabric repository,  I will submit the following related patches:
- https://github.com/satota2/fabric/commit/d6f60009dce9f74caa4df5ac49951d0a0998aac0
- https://github.com/satota2/fabric/commit/fec797b75c40163e9dd92dc1e51acadd4756c7a3
- https://github.com/satota2/fabric/commit/f4d5fff151dc87a597a058184c203b9e9ce72629

#### Related issues
- Main: 
  - https://jira.hyperledger.org/projects/FAB/issues/FAB-17432

- Others: 
  - https://jira.hyperledger.org/projects/FAB/issues/FAB-17401
  - https://jira.hyperledger.org/projects/FAB/issues/FAB-17432
  - https://jira.hyperledger.org/projects/FAB/issues/FAB-17434
  - https://jira.hyperledger.org/projects/FAB/issues/FAB-17435
  - https://jira.hyperledger.org/projects/FAB/issues/FAB-17436

jeffgarratt commented:
Result of behave tests:

Feature: utxo # utxo.feature:11
  As an openchain developer
  I want to be able to launch a 3 peers
1 feature passed, 0 failed, 1 skipped
23 scenarios passed, 0 failed, 4 skipped
369 steps passed, 0 failed, 184 skipped, 0 undefined

Result of go tests:

vagrant@hyperledger-devenv:v0.0.9-4ba4f96:/opt/gopath/src/github.com/hyperledger/fabric/bddtests$ go test -timeout=20m $(go list github.com/hyperledger/fabric/... | grep -v /vendor/ | grep -v /examples/)
?       github.com/hyperledger/fabric/consensus [no test files]
?       github.com/hyperledger/fabric/consensus/controller      [no test files]
?       github.com/hyperledger/fabric/consensus/helper  [no test files]
?       github.com/hyperledger/fabric/consensus/noops   [no test files]
ok      github.com/hyperledger/fabric/consensus/obcpbft 12.557s
ok      github.com/hyperledger/fabric/consensus/statetransfer   0.027s
ok      github.com/hyperledger/fabric/core      0.018s
?       github.com/hyperledger/fabric/core/behave_chaincode/go/table    [no test files]
ok      github.com/hyperledger/fabric/core/chaincode    198.009s
?       github.com/hyperledger/fabric/core/chaincode/platforms  [no test files]
?       github.com/hyperledger/fabric/core/chaincode/platforms/golang   [no test files]
?       github.com/hyperledger/fabric/core/chaincode/shim       [no test files]
?       github.com/hyperledger/fabric/core/chaincode/shim/crypto        [no test files]
ok      github.com/hyperledger/fabric/core/chaincode/shim/crypto/ecdsa  0.275s
?       github.com/hyperledger/fabric/core/config       [no test files]
ok      github.com/hyperledger/fabric/core/container    83.247s
?       github.com/hyperledger/fabric/core/container/ccintf     [no test files]
?       github.com/hyperledger/fabric/core/container/dockercontroller   [no test files]
?       github.com/hyperledger/fabric/core/container/inproccontroller   [no test files]
?       github.com/hyperledger/fabric/core/container/util       [no test files]
ok      github.com/hyperledger/fabric/core/crypto       12.534s
?       github.com/hyperledger/fabric/core/crypto/conf  [no test files]
?       github.com/hyperledger/fabric/core/crypto/ecies [no test files]
ok      github.com/hyperledger/fabric/core/crypto/ecies/generic 0.007s
?       github.com/hyperledger/fabric/core/crypto/utils [no test files]
ok      github.com/hyperledger/fabric/core/db   0.126s
ok      github.com/hyperledger/fabric/core/ledger       27.011s
ok      github.com/hyperledger/fabric/core/ledger/genesis       0.145s
?       github.com/hyperledger/fabric/core/ledger/perfstat      [no test files]
ok      github.com/hyperledger/fabric/core/ledger/statemgmt     0.010s
ok      github.com/hyperledger/fabric/core/ledger/statemgmt/buckettree  0.553s
?       github.com/hyperledger/fabric/core/ledger/statemgmt/raw [no test files]
ok      github.com/hyperledger/fabric/core/ledger/statemgmt/state       1.163s
ok      github.com/hyperledger/fabric/core/ledger/statemgmt/trie        0.321s
?       github.com/hyperledger/fabric/core/ledger/testutil      [no test files]
?       github.com/hyperledger/fabric/core/ledger/util  [no test files]
ok      github.com/hyperledger/fabric/core/peer 0.030s
ok      github.com/hyperledger/fabric/core/rest 0.197s
?       github.com/hyperledger/fabric/core/system_chaincode     [no test files]
?       github.com/hyperledger/fabric/core/system_chaincode/api [no test files]
?       github.com/hyperledger/fabric/core/system_chaincode/sample_syscc        [no test files]
?       github.com/hyperledger/fabric/core/system_chaincode/validity_period_update      [no test files]
ok      github.com/hyperledger/fabric/core/util 2.012s
ok      github.com/hyperledger/fabric/events    0.024s
?       github.com/hyperledger/fabric/events/consumer   [no test files]
?       github.com/hyperledger/fabric/events/producer   [no test files]
?       github.com/hyperledger/fabric/membersrvc        [no test files]
ok      github.com/hyperledger/fabric/membersrvc/ca     0.060s
?       github.com/hyperledger/fabric/membersrvc/protos [no test files]
?       github.com/hyperledger/fabric/peer      [no test files]
ok      github.com/hyperledger/fabric/protos    0.119s

srderson commented:
Also confirmed tests are passing. LGTM.

jyellick commented on core/chaincode/lifecycle/lifecycle.go:
```diff
@@ -496,6 +504,90 @@ func (ef *ExternalFunctions) ApproveChaincodeDefinitionForOrg(chname, ccname str
+		return nil, errors.WithMessagef(err, "could not deserialize namespace metadata for %s", privateName)
+	}
+
+	if ok {
```
In general, I'd prefer to see a:

```
if !ok {
   return nil, errors.Errorf(...)
}
```

rather than deferring to the end of the function.

jyellick commented on core/chaincode/lifecycle/lifecycle.go:
```diff
@@ -496,6 +504,90 @@ func (ef *ExternalFunctions) ApproveChaincodeDefinitionForOrg(chname, ccname str
+	if ok {
+		// Get chaincode parameters for the request sequence
+		ccParameters := &ChaincodeParameters{}
+		if err := ef.Resources.Serializer.Deserialize(NamespacesName, privateName, metadata, ccParameters, orgState); err != nil {
```
Although it should not currently be possible, we need to check the metadata type to ensure that this approval corresponds to a chaincode definition before attempting to deserialize.  Something like:

```
        if metadata.Datatype != ChaincodeDefinitionType {
                return false, nil, errors.Errorf("not a chaincode type: %s", metadata.Datatype)
        }
```

satota2 replied:
The new patch added some logic checking the metadata type.
Instead of ChaincodeDefinition, ApproveChaincodeDefinitionForOrg stores ChaincodeParameters and ChaincodeLocalPackage as the part of the chaincode definition in the Org state. So, the added logic checks the above types.

jyellick commented on core/chaincode/lifecycle/lifecycle.go:
```diff
@@ -496,6 +504,90 @@ func (ef *ExternalFunctions) ApproveChaincodeDefinitionForOrg(chname, ccname str
+		if err != nil {
+			return nil, errors.WithMessagef(err, "could not deserialize chaincode-source metadata for %s", privateName)
+		}
+		if ok {
```
I'd similarly expect a:

```
if !ok {
    return ...
}
```

here.

jyellick reviewed (changes requested):
This generally looks great, thanks for the submission! One thing that needs to be addressed, is checking the metadata type, the rest are just nits.

satota2 committed:
[FAB-17433] Add _lifecycle function for querying approved chaincode

Currently, Fabric peer does not provide a function to query the details of
approved chaincode definitions.

On the other hand, Fabric admins will need to confirm/use the information
after approval for their operations.

This patch adds a function for querying the details of the chaincode
definition approved by the organization itself to `_lifecycle`,
so that Fabric peer can provide a function to query them.

Signed-off-by: Tatsuya Sato <Tatsuya.Sato@hal.hitachi.com>

satota2 commented:
@jyellick
Thank you for your reviews.
I updated the patch to address the comments.

jyellick reviewed (approved):
Looks good to me! Thanks again for the patch, and for addressing the comments.
